feat: Support suffix-based matching for resources

devholic opened the pull request:
**What this PR does / why we need it**:

It would be useful to allow suffix-based matching for resource names, in addition to prefix-based matching.

This commit implements this feature by extending `PrefixWildcard` type instead of using regex since it might be problematic for large numbers of resources as @maxsmythe [commented](https://github.com/open-policy-agent/gatekeeper/issues/1571#issuecomment-925507376) on the issue.

**Which issue(s) this PR fixes**:

Fixes #1571

**Special notes for your reviewer**:

None

maxsmythe commented on pkg/controller/config/process/excluder.go:
```diff
@@ -97,13 +97,13 @@ func (s *Excluder) IsNamespaceExcluded(process Process, obj runtime.Object) (boo
 
 	if obj.GetObjectKind().GroupVersionKind().Kind == "Namespace" && obj.GetObjectKind().GroupVersionKind().Group == "" {
-		return exactOrPrefixMatch(s.excludedNamespaces[process], meta.GetName()), nil
+		return exactOrPrefixSuffixMatch(s.excludedNamespaces[process], meta.GetName()), nil
```
Nit: PrefixSuffixMatch -> WildcardMatch

devholic replied:
Thanks for the review!
I just renamed and rebased onto the latest `master` branch 🙂

maxsmythe reviewed (commented):
Thanks for the PR!

Overall shape looks good, though I didn't have time to finish a close review. One nit that I can find so far.

sozercan commented on pkg/target/target.go:
```diff
@@ -27,20 +27,20 @@ import (
+//     can start with "*" (like `*system`) or "*-" (like `*-system`).
+//   - To implement this, we add the following (\*|\*-)? as a prefix and (\*|-\*)? as a suffix.
+//   - Crucially, this _does not_ allow the value to start or end in a dash (like `-system` or `kube-`).
+//     That is not a valid namespace and not a wildcard, so it's disallowed
```
Does this cover cases like `*-*`? If not, can we call that out too?

devholic replied:
Thanks for pointing this out 🙂

This change doesn't allow more than one asterisk in the pattern, therefore `*-*` is not allowed. I'll add this case to both regexr and comment.

sozercan commented:
Can we update e2e to cover prefix match too? https://github.com/open-policy-agent/gatekeeper/blob/master/test/bats/tests/sync_with_exclusion.yaml#L9

maxsmythe reviewed (approved):
LGTM!

devholic commented:
I rebased onto the latest master branch, and added the following patches:

- Update pattern limitations in comment - efb931a
- Add e2e tests - 5af3a52

sozercan reviewed (approved):
Thank you! LGTM
